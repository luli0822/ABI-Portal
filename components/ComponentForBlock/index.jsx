import React, { useContext } from 'react';
import PropTypes from 'prop-types';
// import { objectExpression, blockStatement } from '@babel/types';
import Hero from '../Hero';
import InlineHero from '../InlineHero';
import TileSection from '../TileSection';
import siteDataContext from '../../contentful/siteDataContext';
/**
 * Map properties of a pageLink or externalLink object to strings for use by the component
 * @param {string} fieldName name of the field
 * @param {object} linkObj link reference object, will be resolved to the actual link object
 * @param {string} lang language for content
 * @param {function} fieldOrFallback fieldOrFallback function that can fetch content from SiteDataContext
 */
function mapLink(fieldName, linkObj, lang, fieldOrFallback) {
	const link = fieldOrFallback(linkObj, lang);
	return {
		[fieldName]: link,
		[`${fieldName}Text`]: fieldOrFallback(link.fields.title, lang),
		[`${fieldName}AccessibleText`]: fieldOrFallback(link.fields.accessibleText, lang),
	};
}

// TODO: consider merging with the useNavigation hook

/**
 * Map properties of an asset object to strings for use by the component
 * @param {string} fieldName name of the field
 * @param {object} linkObj link reference object, will be resolved to the actual media object
 * @param {string} lang language for the content
 * @param {function} fieldOrFallback fieldOrFallback function that can fetch content from SiteDataContext
 */
function mapMedia(fieldName, linkObj, lang, fieldOrFallback) {
	const link = fieldOrFallback(linkObj, lang);
	return {
		[fieldName]: link,
		[`${fieldName}AccessibleText`]: fieldOrFallback(link.fields.description, lang),
	};
}

/**
 * Map properties of a Contentful block object to a simpler, standard props object with content in the current locale.
 * This object is passed to the component, providing content values.
 * @param {object} block The Contentful block object
 * @param {string} lang language for the content
 * @param {function} fieldOrFallback fieldOrFallback function that can fetch content from SiteDataContext
 */
function mapBlockFields(block, lang, fieldOrFallback) {
	// Map + resolve standard fields
	if (!block.fields) return {};

	const id = (block.sys && block.sys.id) || '';
	const resolvedBlocksList = Object.entries(block.fields).map(([fieldName, fieldValue]) => {
		switch (fieldName) {
			case 'blocks':
				return {
					blocks: fieldOrFallback(fieldValue, lang)
						.map(blockRef => fieldOrFallback(blockRef, lang))
						.map(childBlock => mapBlockFields(childBlock, lang, fieldOrFallback)),
				};

			case 'primaryLink':
			case 'secondaryLink':
				return mapLink(fieldName, fieldValue, lang, fieldOrFallback);

			case 'primaryMedia':
			case 'secondaryMedia':
				return mapMedia(fieldName, fieldValue, lang, fieldOrFallback);

			default:
				return { [fieldName]: fieldOrFallback(fieldValue, lang) };
		}
	});

	return Object.assign({}, { id }, ...resolvedBlocksList);
}

/**
 * Given a piece of content from Contentful, picks the appropriate component to render and passes the
 * simplified, standardized content props for the current locale to that component.
 * @param {object} props
 * @param {object} props.block The block reference link for the content to pass to the selected component
 * @param {boolean} props.isFirstChild Indicate if this is the first component on the page, used to determine where to render H1 tag
 */
function ComponentForBlock({ block, isFirstChild }) {
	let Tag = InlineHero;

	const { fieldOrFallback, lang } = useContext(siteDataContext);

	// resolve the data from a link to actual entry
	const blockValue = fieldOrFallback(block);
	const content = mapBlockFields(blockValue, lang, fieldOrFallback);
	const displayAs = fieldOrFallback(blockValue.fields.displayAs);

	const commonProps = {
		content,
		isFirstChild,
	};

	const componentProps = {};

	switch (blockValue.sys.contentType.sys.id) {
		// TODO - consider making this reference a dictionary of "display-as" values,
		// with fallbacks for complexBlock and blockGroup
		// instead of a big switch/case tree
		case 'complexBlock':
			switch (displayAs) {
				case 'inline-hero':
					Tag = InlineHero;
					break;
				default:
					Tag = Hero;
					break;
			}
			break;

		case 'blockGroup':
			Tag = TileSection;
			break;

		default:
			break;
	}

	// eslint-disable-next-line react/jsx-props-no-spreading
	return <Tag {...commonProps} {...componentProps} />;
}

ComponentForBlock.propTypes = {
	block: PropTypes.shape({
		sys: PropTypes.shape({
			id: PropTypes.string.isRequired,
			type: PropTypes.string.isRequired,
			linkType: PropTypes.string.isRequired,
		}).isRequired,
	}).isRequired,
	isFirstChild: PropTypes.bool.isRequired,
};

export default ComponentForBlock;
