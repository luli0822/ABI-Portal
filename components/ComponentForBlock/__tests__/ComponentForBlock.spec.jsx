import React from 'react';
import { shallow } from 'enzyme';
import ComponentForBlock from '../index';
import mockPushDataLayer from '../../../contentful/__mocks__/analyticsContext';

jest.mock('react', () => {
	const linkedEntries = {
		'page-link-id': {
			sys: {
				id: 'page-link-id',
				contentType: {
					sys: {
						id: 'pageLink',
					},
				},
			},
			fields: {
				title: {
					'en-CA': 'Page Link Title EN',
				},
				accessibleText: {
					'en-CA': 'Page Link Accessible Text EN',
				},
			},
		},
		'complex-block-id': {
			fields: {
				title: {
					'en-CA': 'Complex Block Title EN',
				},
				primaryLink: {
					'en-CA': {
						sys: {
							id: 'page-link-id',
							type: 'Link',
							linkType: 'Entry',
						},
					},
				},
			},
			sys: {
				id: 'complex-block-id',
				contentType: {
					sys: {
						id: 'complexBlock',
					},
				},
			},
		},
		'block-group-id': {
			fields: {
				title: {
					'en-CA': 'Complex Block Title EN',
				},
				blocks: {
					'en-CA': [
						{
							sys: {
								id: 'complex-block-id',
								type: 'Link',
								linkType: 'Entry',
							},
						},
					],
				},
			},
			sys: {
				id: 'block-group-id',
				contentType: {
					sys: {
						id: 'blockGroup',
					},
				},
			},
		},
	};

	const linkedAssets = {};

	// TODO - find better way to mock fieldOrFallback - this is just copy/paste from <SiteData />
	function isLink(field) {
		return field && field.sys && field.sys.type === 'Link';
	}

	function fieldOrFallback(field, fieldLang) {
		if (isLink(field)) {
			if (field.sys.linkType === 'Entry') return linkedEntries[field.sys.id];
			if (field.sys.linkType === 'Asset') return linkedAssets[field.sys.id];
		}

		const fieldValue = field ? field[fieldLang] || field['en-CA'] : undefined;

		if (isLink(fieldValue)) {
			if (fieldValue.sys.linkType === 'Entry') return linkedEntries[fieldValue.sys.id];
			if (fieldValue.sys.linkType === 'Asset') return linkedAssets[fieldValue.sys.id];
		}

		return fieldValue;
	}

	const ActualReact = require.requireActual('react');

	return {
		...ActualReact,
		useContext: () => ({
			fieldOrFallback,
			lang: 'en-CA',
			pushDataLayer: mockPushDataLayer,
		}),
	};
});

const complexBlockProps = {
	block: {
		sys: {
			id: 'complex-block-id',
			type: 'Link',
			linkType: 'Entry',
		},
	},
	isFirstChild: true,
};

const blockGroupProps = {
	block: {
		sys: {
			id: 'block-group-id',
			type: 'Link',
			linkType: 'Entry',
		},
	},
	isFirstChild: true,
};

describe('<ComponentForBlock />', () => {
	it('renders Complex Block', () => {
		// eslint-disable-next-line react/jsx-props-no-spreading
		const component = shallow(<ComponentForBlock {...complexBlockProps} />);
		expect(component.debug()).toMatchSnapshot();
	});
	it('renders correct props for Complex Block', () => {
		// eslint-disable-next-line react/jsx-props-no-spreading
		const component = shallow(<ComponentForBlock {...complexBlockProps} />);
		expect(component.props()).toMatchSnapshot();
	});
	it('renders Block Group', () => {
		// eslint-disable-next-line react/jsx-props-no-spreading
		expect(shallow(<ComponentForBlock {...blockGroupProps} />).debug()).toMatchSnapshot();
	});
	it('renders correct props for Block Group', () => {
		// eslint-disable-next-line react/jsx-props-no-spreading
		const component = shallow(<ComponentForBlock {...blockGroupProps} />);
		expect(component.props()).toMatchSnapshot();
	});
});
