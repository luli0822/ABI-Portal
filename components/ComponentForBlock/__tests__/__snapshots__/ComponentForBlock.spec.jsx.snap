// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComponentForBlock /> renders Block Group 1`] = `"<TileSection content={{...}} isFirstChild={true} />"`;

exports[`<ComponentForBlock /> renders Complex Block 1`] = `"<Hero content={{...}} isFirstChild={true} />"`;

exports[`<ComponentForBlock /> renders correct props for Block Group 1`] = `
Object {
  "content": Object {
    "blocks": Array [
      Object {
        "id": "complex-block-id",
        "primaryLink": Object {
          "fields": Object {
            "accessibleText": Object {
              "en-CA": "Page Link Accessible Text EN",
            },
            "title": Object {
              "en-CA": "Page Link Title EN",
            },
          },
          "sys": Object {
            "contentType": Object {
              "sys": Object {
                "id": "pageLink",
              },
            },
            "id": "page-link-id",
          },
        },
        "primaryLinkAccessibleText": "Page Link Accessible Text EN",
        "primaryLinkText": "Page Link Title EN",
        "title": "Complex Block Title EN",
      },
    ],
    "id": "block-group-id",
    "title": "Complex Block Title EN",
  },
  "isFirstChild": true,
}
`;

exports[`<ComponentForBlock /> renders correct props for Complex Block 1`] = `
Object {
  "content": Object {
    "id": "complex-block-id",
    "primaryLink": Object {
      "fields": Object {
        "accessibleText": Object {
          "en-CA": "Page Link Accessible Text EN",
        },
        "title": Object {
          "en-CA": "Page Link Title EN",
        },
      },
      "sys": Object {
        "contentType": Object {
          "sys": Object {
            "id": "pageLink",
          },
        },
        "id": "page-link-id",
      },
    },
    "primaryLinkAccessibleText": "Page Link Accessible Text EN",
    "primaryLinkText": "Page Link Title EN",
    "title": "Complex Block Title EN",
  },
  "isFirstChild": true,
}
`;
